fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownPredicateScanRule.java

```diff
@@ -98,6 +103,33 @@ public List<OptExpression> transform(OptExpression input, OptimizerContext conte
         predicates = scalarOperatorRewriter.rewrite(predicates,
                 ScalarOperatorRewriter.DEFAULT_REWRITE_SCAN_PREDICATE_RULES);
         predicates = Utils.transTrue2Null(predicates);
+        if (context.getSessionVariable().isEnableScanPredicateExprReuse()) {
+            OlapTablePredicateExtractor olapTablePredicateExtractor =
+                    new OlapTablePredicateExtractor(logicalScanOperator.getColRefToColumnMetaMap());
+            olapTablePredicateExtractor.extract(predicates);
+            ScalarOperator pushedPredicates = olapTablePredicateExtractor.getPushedPredicates();
+            ScalarOperator reservedPredicates = olapTablePredicateExtractor.getReservedPredicates();
+            boolean newScanPredicateIsSame = Objects.equals(pushedPredicates, scanPredicate);
+            boolean newFilterPredicateIsSame = Objects.equals(filterPredicate, reservedPredicates);
+            // if nothing changed after extracting predicates, just return
+            if (newScanPredicateIsSame && newFilterPredicateIsSame) {
+                return Lists.newArrayList();
+            }
+            if (reservedPredicates != null) {
+                Operator.Builder builder = OperatorBuilderFactory.build(logicalScanOperator);
+                LogicalScanOperator newScanOperator = (LogicalScanOperator) builder.withOperator(logicalScanOperator)
+                           .setPredicate(pushedPredicates).build();
+                newScanOperator.buildColumnFilters(pushedPredicates);
+                Map<ColumnRefOperator, ScalarOperator> projectMap =
+                        newScanOperator.getOutputColumns().stream()
+                                .collect(Collectors.toMap(Function.identity(), Function.identity()));
+                LogicalProjectOperator logicalProjectOperator = new LogicalProjectOperator(projectMap);
```

Comment:
is this project operator necessary?

Reply:
> is this project operator necessary?

Just to be consistent with the previous, I don't want to handle the projection columns specifically here because other rules can prune it if necessary.